Xamarin.Forms.Platform.iOS/Renderers/ListViewRenderer.cs

```diff
@@ -692,22 +671,42 @@ internal nfloat GetEstimatedRowHeight(UITableView table)
 				return CalculateHeightForCell(table, firstCell);
 			}
 
-			public override nfloat EstimatedHeight(UITableView tableView, NSIndexPath indexPath)
+			internal override void InvalidatePrototypicalCellCache()
+			{
+				_prototypicalCellByTypeOrDataTemplate.Clear();
+			}
+
+			internal Cell GetPrototypicalCell(NSIndexPath indexPath)
 			{
-				if (_useEstimatedRowHeight)
-					return tableView.EstimatedRowHeight;
+				var itemTypeOrDataTemplate = default(object);
+
+				var cachingStrategy = List.CachingStrategy;
+				if (cachingStrategy == ListViewCachingStrategy.RecycleElement)
+					itemTypeOrDataTemplate = GetDataTemplateForPath(indexPath);
```

Comment:
This is where the existing Recycle strategy is improved at the cost of a slight break whereby a DataTemplate can no longer return a different type of cell. Requiring a DataTemplate always returns the same type of cell allows us to use the DataTemplate to hit a cache here to get a prototypical cell for height measuring. It's more expensive than the new recycle as we always have to ask for the DataTemplate for an item instead just using the item's type as the cache key.